### ISSUE TITLE
[v2] Emacs 29+ check for executable prevents Treesitter setup

### ISSUE THREAD

**Issue by jvdydev** [SystemCrafters/crafted-emacs#363]
Emacs Versions affected: 29+
Tested on: Arch Linux, Emacs 29 custom w/ treesitter + native-comp

While tinkering with the RC1, I realized that treesitter is installed as a link library (at least on Arch), therefor the check for loading `treesit-auto` always returns nil for `(executable-find "tree-sitter")`.

How to reproduce (roughly):
- acquire Emacs 29 with treesitter enabled
- Configure `crafted-ide` module
- Run `M-: (featurep 'treesit-auto)`

Expected Result: `t`

Actual Result: `nil`

Proposal for fix: Remove the `(executable-find ...)` call
```emacs-lisp
;; before
(when (version< "29" emacs-version)
  (when (and (member "TREE_SITTER" (split-string system-configuration-features))
             (executable-find "tree-sitter"))
    ...))

;; after
(when (version< "29" emacs-version)
  (when (member "TREE_SITTER" (split-string system-configuration-features))
    ...))
```

**Comment by sthesing** [comment]
> Proposal for fix: Remove the `(executable-find ...)` call

Hm, yeah, under the circumstances that seems to be the best option to make this work as intended.

**Comment by jvdydev** [comment]
The problem is that I have no idea how this would behave for e.g. Emacs w/ treesitter in an environment where there is no treesitter library.
I also am not sure what platform/distribution actually has a "tree-sitter" executable.

While I did propose the change, I'm not sure it's the correct one to make here.
I may need to test more distros and see how it behaves with/without treesitter installed (and potential workarounds).

**Comment by jvdydev** [comment]
Also `featurep` would reproduce this as the `(require 'treesit-auto)` never gets run (on a system with the treesitter lib), so the library is not loaded, hence `nil` is returned.

Once `(require 'treesit-auto)` is ran, the library is loaded, therefor `featurep` returns `t`.

`executable-find` searches for an executable in `$PATH`.

**Comment by sthesing** [comment]
> The problem is that I have no idea how this would behave for e.g. Emacs w/ treesitter in an environment where there is no treesitter library. I also am not sure what platform/distribution actually has a "tree-sitter" executable.

I would guess that - just as Arch does it - sensible GNU/Linux distributions will list treesitter as a dependency for Emacs that has ts support. Windows might become an issue, though.

And a whole other topic is the question whether an executable will be there at all and if it will be called `tree-sitter`, `treesitter` or something else entirely.
All the more reason not to look for one, I think.

**Comment by jeffbowman** [comment]
I'm about to travel for 5 hours, but I'll look more into this when I get there (short weekend away from home).  Initial question would be maybe switch to using the `require` guard?  But the question still remains: what happens when `tree-sitter` is _not_ installed on the host system?  I may need to spin up a VM (or a few) to test different scenarios like this.  I "think" when I originally put this together there was such a command as `tree-sitter`, but that is past history.  I've turned off `tree-sitter` in my configuration as I was having difficulties with it, but didn't have time to go back and debug why.  Might be nice to figure that out. 

@jvdydev if you figure something out in the meanwhile, I'm very curious :-)

**Comment by jvdydev** [comment]
ye, I'll spin up a few distro containers and test the `crafted-ide` module.

`treesit-auto` already is in a require guard (directly following the outer guard in question with the `executable-find`).
(There is some more code for `combobulate` in the outer guard which is equally affected as `treesit-auto`).
I have no idea how those two behave if treesitter is not installed and they are loaded.

I think the problem is we don't have any easy way (to my knowledge) to check if an OS-level library (so/dll/dylib) is installed.
`treesit` has a `treesit-available-p`, I'll see if that detects availability of the library (it's defined in C source, so maybe it can).

To summarize what we roughly are looking for:
```emacs-lisp
(when (magic-expr-treesitter-is-fully-set-up-p) ; fully => OS-level + Emacs features
  (when (require 'treesit-auto nil :noerror)
    ;; configure treesit-auto
    )
  (when (locate-library "combobulate")
    ;; configure combobulate
    ))
```

The tl;dr, I'll try to make a checklist of combinations to test and test them on at least the major distros (Arch, Debian, Fedora), we'll see from there.

**Comment by jvdydev** [comment]
I also just realized with a quick test that we have a lot less concerns.

Turns out Emacs with treesitter enabled doesn't start unless you have libtree-sitter (I quickly uninstalled treesitter on my machine and tried to reboot Emacs):
```
emacs: error while loading shared libraries: libtree-sitter.so.0: cannot open shared object file: No such file or directory
```

**Comment by jvdydev** [comment]
Meaning if Emacs *is* configured with treesitter, the library is there (or it probably won't boot).

Means the original proposal would be fine?

**Comment by sthesing** [comment]
Guess so. If we're only there to check for that library if it is present in the first place, then we don't need to check.
Thanks for looking into this!

**Comment by jeffbowman** [comment]
That's interesting.  So maybe we just need to check if the feature is enabled?

Combobulate and treesit-auto throw warnings when a mode is started, but the appropriate language ts module is not available. Which is the point of using ts auto - to automatically install and load the appropriate dll/so/dylib.

**Comment by jeffbowman** [comment]
Ah... And I see there is a PR along those lines... 😂

**Comment by jvdydev** [comment]
Just an update to this issue too, the PR works on Windows (and Linux).

Since tree-sitter needs to compile the language grammars locally, a C compiler (more specifically, at least one of `'("cc" "gcc" "c99")` for Emacs' `tree-sitter`) is required to be available through `$PATH` (or `$env:Path` on Windows).

There may be a place for documenting this and/or not loading these packages unless there is a C compiler there, but that's up for further discussion (preferably in a new issue, as it is not related to the `executable-find` issue).
